micropip/package_index.py

```diff
@@ -303,36 +308,101 @@ async def query_package(
 
     index_urls = [PYPI_URL if url == PYPI else url for url in index_urls]
 
-    for url in index_urls:
-        logger.debug("Looping through index urls: %r", url)
-        if _contain_placeholder(url):
-            url = url.format(package_name=name)
-            logger.debug("Formatting url with package name : %r", url)
-        else:
-            url = f"{url}/{name}/"
-            logger.debug("Url has no placeholder, appending package name : %r", url)
-        try:
-            metadata, headers = await fetch_string_and_headers(url, _fetch_kwargs)
-        except HttpStatusError as e:
-            if e.status_code == 404:
-                logger.debug("NotFound (404) for %r, trying next index.", url)
+    projects_info = []
+
+    # With "first-index" strategy, we'll return the first match we find
+    # without checking the other indexes at all.
+    if strategy == "first-index":
```

Comment:
Each strategy has duplicate logics, so I think we can separate this with multiple functions, extracting out common parts.